File: drivers/io/targets/at91sam4s/io.c
#define __private_include__
#include <flipper/io/io.h>

#include <platform.h>

void io_configure(void) {

	/* ~ Enable the PIOA (Peripheral IO Controller A) in the PCER (Peripheral Clock Enable Register). ~ */
	set_bit_in_port(AT91C_ID_PIOA, AT91C_BASE_PMC -> PMC_PCER);

}

void io_set_direction(uint8_t pin, uint8_t direction) {

	/* ~ If the pin is less than or equal to IO16, the pin is digital. ~ */
	if (pin <= IO16)

		digital_set_direction(pin, direction);

	/* ~ If not, the pin is analog. ~ */
	else

		analog_set_direction(pin, direction);

}

void io_write(uint8_t pin, uint16_t value) {

	/* ~ If the pin is less than or equal to IO16, the pin is digital. ~ */
	if (pin <= IO16)

		digital_write(pin, value);

	/* ~ If not, the pin is analog. ~ */
	else

		analog_write(pin, value);

}

uint16_t io_read(uint8_t pin) {

	/* ~ If the pin is less than or equal to IO16, the pin is digital. ~ */
	if (pin <= IO16)

		return digital_read(pin);

	/* ~ If not, the pin is analog. ~ */
	else

		return analog_read(pin);

}

File: languages/java/src/Io.java
/**
 * Created by Nick Mosher on 4/1/16.
 */
public class Io {

    private Flipper._flipper mFlipper;

    public Io(Flipper._flipper flipper) {
        mFlipper = flipper;
    }

    public void configure() {
        mFlipper.io_configure();
    }

    /**
     * Sets the direction of this pin. True means output, false means input.
     * @param direction The direction (input or output) for this pin.
     */
    public void direction(int pin, boolean direction) {
        mFlipper.io_set_direction((byte) pin, direction ? (byte) 1 : (byte) 0);
    }

    /**
     * Write a boolean value to this IO output.
     * @param value The boolean value to write.
     */
    public void write(int pin, boolean value) {
        if(pin < 0 || pin > 16) {
            throw new IllegalArgumentException("Cannot write boolean value to analog pin.");
        }

        mFlipper.io_write((byte) pin, value ? (byte) 1 : (byte) 0);
    }

    /**
     * Write an analog value to this IO output.
     * @param value The analog value to write.
     */
    public void write(int pin, short value) {
        if(pin < 17) {
            throw new IllegalArgumentException("Cannot write analog value to digital pin.");
        }

        mFlipper.io_write((byte) pin, value);
    }

    public boolean digitalRead(int pin) {
        if(pin < 0 || pin > 16) {
            throw new IllegalArgumentException("Cannot read boolean value from analog pin.");
        }

        return mFlipper.io_read((byte) pin) != 0;
    }

    public short analogRead(int pin) {
        if(pin < 17) {
            throw new IllegalArgumentException("Cannot read analog value from digital pin.");
        }

        return mFlipper.io_read((byte) pin);
    }
}
